### PR TITLE
feat(deployer): send runtime logs to the logger service

### PR DESCRIPTION
**Pull request by orhun**
Simply capture logs and send them to the logger service.

yolo.

### CODE REVIEW

File: deployer/src/runtime_manager.rs [hunk 1]
```diff
@@ -15,7 +19,7 @@ type Runtimes = Arc<
         HashMap<
             Uuid,
             (
-                process::Child,
+                Arc<Mutex<process::Child>>,
```
Comment (iulianbarbu) [inline deployer/src/runtime_manager.rs:22]: Do we need the `Arc<Mutex<..` here? I think `Child` is `Send` and `Sync` already [1], [2].

[1] https://docs.rs/tokio/latest/tokio/process/struct.Child.html#impl-Send-for-Child
[2] https://docs.rs/tokio/latest/tokio/process/struct.Child.html#impl-Sync-for-Child

File: deployer/src/runtime_manager.rs [hunk 2]
```diff
@@ -115,19 +131,56 @@ impl RuntimeManager {
...
+        // Ensure the child process is spawned in the runtime so it can
+        // make progress on its own while we await for any output.
+        tokio::spawn(async move {
+            process_cloned
+                .lock()
+                .await
+                .wait()
+                .await
+                .expect("child process encountered an error");
+        });
```
Comment (iulianbarbu) [inline deployer/src/runtime_manager.rs:157]: Just wondering, is it worth waiting in a task that isn't awaited?

File: deployer/src/runtime_manager.rs [hunk 3]
```diff
@@ -115,19 +131,56 @@ impl RuntimeManager {
...
+
+        let mut reader = BufReader::new(stdout).lines();
+        let mut logger_client = self.logger_client.clone();
+        tokio::spawn(async move {
+            while let Some(line) = reader.next_line().await.unwrap() {
+                // TODO: `store_logs` accepts a Vec but logs are sent one by one. Is this feasible?
```
Comment (iulianbarbu) [inline deployer/src/runtime_manager.rs:163]: I think we can probably start by adding some default batching logic here. It should account lets say for batching 100 logs one time, but it should also account for some time window. E.g. whenever deciding if we want to store the logs, we would verify the batch size against the default value or check whether the window has passed.

File: cargo-shuttle/src/lib.rs [hunk 4]
```diff
@@ -641,6 +641,7 @@ impl Shuttle {
         };
 
         // Child process and gRPC client for sending requests to it
+        // TODO: Is child stdout should be handled?
```
Comment (chesedo) [inline cargo-shuttle/src/lib.rs:644]: This should just print to stdout

File: deployer/src/runtime_manager.rs (hunk 3 above)
Comment (chesedo) [inline deployer/src/runtime_manager.rs:163]: Correct, we'll make another PR to bundle them somehow

**Review by chesedo — COMMENTED** [review]
For the CLI, we just have to print the logs to stdout too :smile:

File: deployer/src/runtime_manager.rs [hunk 5]
```diff
@@ -115,19 +131,56 @@ impl RuntimeManager {
...
+                .lock()
+                .await
+                .wait()
+                .await
+                .expect("child process encountered an error");
+        });
+
+        let mut reader = BufReader::new(stdout).lines();
+        let mut logger_client = self.logger_client.clone();
+        tokio::spawn(async move {
```
Comment (iulianbarbu) [inline deployer/src/runtime_manager.rs:161]: I am wondering if it makes sense to handle this logging related task in the outer scope. Probably it means propagating this up to `task` from `run.rs`.

File: deployer/src/runtime_manager.rs [hunk 6]
```diff
@@ -115,19 +131,56 @@ impl RuntimeManager {
...
+            while let Some(line) = reader.next_line().await.unwrap() {
+                // TODO: `store_logs` accepts a Vec but logs are sent one by one. Is this feasible?
+                let logs = vec![LogItem {
+                    deployment_id: id.to_string(),
+                    log_line: Some(LogLine {
+                        service_name: service_name.to_string(),
+                        tx_timestamp: Some(Timestamp::from(SystemTime::UNIX_EPOCH)),
```
Comment (iulianbarbu) [inline deployer/src/runtime_manager.rs:168]: We'll need to use here something like `Utc::now()`.

File: deployer/src/runtime_manager.rs (hunk 1 above)
Comment (orhun) [reply deployer/src/runtime_manager.rs:22]: Yeah, but that's not the reason why I used `Mutex`.

See: https://github.com/orhun/shuttle/blob/1370efaaaf68f774094c180842d69422f0202355/deployer/src/runtime_manager.rs#L134-L140

Here, `process` is being moved to `self.runtimes` and we need to use it mutably afterwards in another closure (`tokio::spawn`) so there isn't a way to do that without an async `Mutex` and atomic reference counter since it \*does not\* implement `Clone`.

File: deployer/src/runtime_manager.rs (hunk 2 above)
Comment (orhun) [reply deployer/src/runtime_manager.rs:157]: From the docs:

> The provided future will start running in the background immediately when spawn is called, even if you don’t await the returned `JoinHandle`.

File: deployer/src/runtime_manager.rs (hunk 6 above)
Comment (orhun) [reply deployer/src/runtime_manager.rs]: f647dab028b08aa76898aec9bb136ea6116003cb

File: deployer/src/runtime_manager.rs (hunk 5 above)
Comment (chesedo) [reply deployer/src/runtime_manager.rs:162]: `run.rs` might not have access to the logger client though

Comment (iulianbarbu) [reply deployer/src/runtime_manager.rs:162]: I meant just propagating the task handle, no need for the client in that scope. In some cases logging sending might break so detecting clearly when that happens should help.

File: deployer/src/runtime_manager.rs (hunk 6 above)
Comment (iulianbarbu) [reply deployer/src/runtime_manager.rs]: Would be great though to add something like this for nanos: https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.timestamp_subsec_nanos.

File: cargo-shuttle/src/lib.rs (hunk 4 above)
Comment (orhun) [reply cargo-shuttle/src/lib.rs:644]: I'm getting the following error for `cargo shuttle run`:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: runtime client endpoint did not open in time
```

Still investigating, do you have any ideas?

File: deployer/src/runtime_manager.rs (hunk 3 above)
Comment (iulianbarbu) [reply deployer/src/runtime_manager.rs:164]: Agreed. Resolving this.

File: cargo-shuttle/src/lib.rs (hunk 4 above)
Comment (orhun) [reply cargo-shuttle/src/lib.rs:644]: Ah, I think it is because I'm not using `git` dependencies for my test project.

File: deployer/src/runtime_manager.rs (hunk 1 above)
Comment (iulianbarbu) [reply deployer/src/runtime_manager.rs:23]: Is this task where the process is moved the one which calls `.wait()` on it. I don't find it necessarily useful, but lmk if you disagree.

File: cargo-shuttle/src/lib.rs (hunk 4 above)
Comment (orhun) [reply cargo-shuttle/src/lib.rs]: da38f862bcce4520f213673e51d2d451c5201f78

File: deployer/src/runtime_manager.rs (hunk 1 above)
Comment (orhun) [reply deployer/src/runtime_manager.rs]: Yeah, I tested in da38f862bcce4520f213673e51d2d451c5201f78 and it worked fine without that.

Removed in 002d0c01a408f49bfd49bbe4a82f042bdb6b5847

File: deployer/src/runtime_manager.rs (hunk 2 above)
Comment (orhun) [reply deployer/src/runtime_manager.rs]: Yup, 002d0c01a408f49bfd49bbe4a82f042bdb6b5847

File: cargo-shuttle/src/lib.rs [hunk 7]
```diff
@@ -658,6 +658,18 @@ impl Shuttle {
...
 
+        let child_stdout = runtime
+            .stdout
+            .take()
+            .context("child process did not have a handle to stdout")?;
+        let mut reader = BufReader::new(child_stdout).lines();
+        tokio::spawn(async move {
+            while let Some(line) = reader.next_line().await.unwrap() {
+                let mut stdout = stdout().lock();
+                writeln!(stdout, "{}", line).unwrap();
```
Comment (iulianbarbu) [inline cargo-shuttle/src/lib.rs:669]: Why not use `println!`?

File: deployer/src/runtime_manager.rs (hunk 6 above)
Comment (orhun) [reply deployer/src/runtime_manager.rs]: f2d42cbda13a34558f238b8a6f97c769b51af2d0

File: cargo-shuttle/src/lib.rs (hunk 7 above)
Comment (orhun) [reply cargo-shuttle/src/lib.rs:669]: From `println` docs:

> The println! macro will lock the standard output on each call. If you call println! within a hot loop, this behavior may be the bottleneck of the loop. To avoid this, lock stdout with [io::stdout().lock()](https://doc.rust-lang.org/std/io/struct.Stdout.html)

https://doc.rust-lang.org/std/macro.println.html

Also, I believe using `stdout` makes the implementation more explicit.

Comment (iulianbarbu) [reply cargo-shuttle/src/lib.rs:669]: Ah, interesting. However, wouldn't the locking happening at each call of `println!` be the same as locking on `stdout inside the loop every time before the `writeln!`? I think the idea from the doc is to `lock` outside the loop and use the guard to write continuously to the stdout - meaning lock once for the entirety of the loop instead of locking on each iteration.

Comment (iulianbarbu) [reply cargo-shuttle/src/lib.rs:669]: >  I think the idea from the doc is to lock outside the loop and use the guard to write continuously to the stdout - meaning lock once for the entirety of the loop instead of locking on each iteration.

However, thinking better about this, we might need to write to the stdout of the cargo-shuttle process from different tasks too. I think this would block the other tasks.

Comment (jonaro00) [reply cargo-shuttle/src/lib.rs:669]: The speed of this loop is not super important either.

Comment (iulianbarbu) [reply cargo-shuttle/src/lib.rs:669]: @orhun what are your thoughts on this? Do you still find the `stdout` locking approach useful? We might be able to close this, I think this is the last unresolved comment.

Comment (orhun) [reply cargo-shuttle/src/lib.rs:669]: I'm okay with either way. If you're fine with `stdout` approach, feel free to give it a final review and we can proceed!

Comment (iulianbarbu) [reply cargo-shuttle/src/lib.rs:669]: Not sure what you mean by the final review (I did review the `stdout` approach), but I would remove the `stdout` and replace it with `println!` just because using it like you did they should be the same from my understanding. Correct me if I am wrong.

Comment (orhun) [reply cargo-shuttle/src/lib.rs]: 14d7dfe27ae54d0d92a01d971bad8e884ff06dc5

File: deployer/src/deployment/deploy_layer.rs [hunk 8]
```diff
@@ -449,10 +454,24 @@ mod tests {
...
 
+        let channel = Endpoint::try_from(logger_uri.to_string())
+            .unwrap()
+            .connect()
+            .await
+            .expect("failed to connect to logger");
+
+        let channel = ServiceBuilder::new()
+            .layer(ClaimLayer)
+            .layer(InjectPropagationLayer)
```
Comment (chesedo) [inline deployer/src/deployment/deploy_layer.rs:465]: Hmm, we should be fine without these layers in the tests.

File: deployer/src/deployment/run.rs [hunk 9]
```diff
@@ -526,10 +538,24 @@ mod tests {
...
 
+        let channel = Endpoint::try_from(logger_uri.to_string())
+            .unwrap()
+            .connect()
+            .await
+            .expect("failed to connect to logger");
+
+        let channel = ServiceBuilder::new()
+            .layer(ClaimLayer)
+            .layer(InjectPropagationLayer)
```
Comment (chesedo) [inline deployer/src/deployment/run.rs:549]: Here too

**Review by chesedo — COMMENTED** [review]
Just simplification notes

File: deployer/src/deployment/deploy_layer.rs (hunk 8 above)
Comment (orhun) [reply deployer/src/deployment/deploy_layer.rs]: e2edeac07a50b5166c7f2fb1c8332a65020bc2d4

File: deployer/src/deployment/run.rs (hunk 9 above)
Comment (orhun) [reply deployer/src/deployment/run.rs]: e2edeac07a50b5166c7f2fb1c8332a65020bc2d4

File: deployer/src/deployment/deploy_layer.rs (hunk 8 above)
Comment (oddgrd) [reply deployer/src/deployment/deploy_layer.rs]: The runtime manager expects a client with those layers, so we'll need to keep them (or refactor the runtime manager to create the client instead). I think we discussed removing the claim layer too, right @chesedo?

Comment (chesedo) [reply deployer/src/deployment/deploy_layer.rs]: Actually the server wants those layers. Since we will mock / stub the server, we can control the layers needed in each test

**Review by chesedo — APPROVED** [review]
LGTM
